csrc/ec_gen.cpp

```diff
@@ -183,16 +180,29 @@ JNIEXPORT void JNICALL Java_com_amazon_corretto_crypto_provider_EcGen_generateEc
         // Now that we have the params, extract the group from them (all we care about)
         CHECK_OPENSSL(group = EC_KEY_get0_group(ecParams));
 
-        // Build the structure which will hold our result
-        if(!likely(key = EC_KEY_new())) {
-            throw java_ex(EX_OOM, "Out of memory");
+        // Setup the temporary PKEY context with the extracted parameters.
+        if (!paramCtx.setKeyCtx(EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL))) {
+            throw_openssl("Unable to create new evp key context");
         }
-
-        CHECK_OPENSSL(EC_KEY_set_group(key, group));
-
-        if (!likely(EC_KEY_generate_key(key))) {
-            throw_openssl("Unable to generate key");
+        CHECK_OPENSSL(EVP_PKEY_paramgen_init(paramCtx.getKeyCtx()) == 1);
+        CHECK_OPENSSL(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(paramCtx.getKeyCtx(), EC_GROUP_get_curve_name(group)) == 1);
+        // We need to define a temporary pointer to an EVP_PKEY
+        // to be able to call EVP_PKEY_paramgen function.
+        EVP_PKEY *tmpKey0 = NULL;
+        CHECK_OPENSSL(EVP_PKEY_paramgen(paramCtx.getKeyCtx(), &tmpKey0) == 1);
+        paramCtx.setKey(tmpKey0);
+
+        // Generate the actual EC key.
+        if (!keyCtx.setKeyCtx(EVP_PKEY_CTX_new(paramCtx.getKey(), NULL))) {
+            throw_openssl("Unable to create new evp key context");
         }
+        CHECK_OPENSSL(EVP_PKEY_keygen_init(keyCtx.getKeyCtx()) == 1);
+        // We need to define a temporary pointer to an EVP_PKEY
+        // to be able to call EVP_PKEY_keygen function.
+        EVP_PKEY *tmpKey1 = NULL;
+        CHECK_OPENSSL(EVP_PKEY_keygen(keyCtx.getKeyCtx(), &tmpKey1) == 1);
+        keyCtx.setKey(tmpKey1);
```

Comment:
why do we not need to free `tmpKey0` or `tmpKey1`? does `setKey` transfer ownership to the EvpKeyContext struct?

Reply:
yes, I think that's the reason Greg prefers to use `EvpKeyContext`